src/main/java/com/android/volley/cronet/CronetHttpStack.java

```diff
@@ -0,0 +1,338 @@
+                    builder.setHttpMethod("POST");
+                    addBodyIfExists(postBody, builder);
+                } else {
+                    builder.setHttpMethod("GET");
+                }
+                break;
+            case Request.Method.GET:
+                // Not necessary to set the request method because connection defaults to GET but
+                // being explicit here.
+                builder.setHttpMethod("GET");
+                break;
+            case Request.Method.DELETE:
+                builder.setHttpMethod("DELETE");
+                break;
+            case Request.Method.POST:
+                builder.setHttpMethod("POST");
+                addBodyIfExists(request.getBody(), builder);
+                break;
+            case Request.Method.PUT:
+                builder.setHttpMethod("PUT");
+                addBodyIfExists(request.getBody(), builder);
+                break;
+            case Request.Method.HEAD:
+                builder.setHttpMethod("HEAD");
+                break;
+            case Request.Method.OPTIONS:
+                builder.setHttpMethod("OPTIONS");
+                break;
+            case Request.Method.TRACE:
+                builder.setHttpMethod("TRACE");
+                break;
+            case Request.Method.PATCH:
+                builder.setHttpMethod("PATCH");
+                addBodyIfExists(request.getBody(), builder);
+                break;
+            default:
+                throw new IllegalStateException("Unknown method type.");
+        }
+    }
+
+    /**
+     * Sets the request headers for the UrlRequest.
+     *
+     * @param request to get the headers from
+     * @param additionalHeaders for the UrlRequest
+     * @param builder that we are adding the request headers to
+     * @throws AuthFailureError is thrown if Request#getHeaders throws ones
+     */
+    private void setRequestHeaders(
+            Request<?> request, Map<String, String> additionalHeaders, UrlRequest.Builder builder)
+            throws AuthFailureError {
+        HashMap<String, String> map = new HashMap<>();
+        map.putAll(additionalHeaders);
+        // Request.getHeaders() takes precedence over the given additional (cache) headers).
+        map.putAll(request.getHeaders());
+        for (Map.Entry<String, String> header : map.entrySet()) {
+            builder.addHeader(header.getKey(), header.getValue());
+        }
+    }
+
+    /** Sets the UploadDataProvider of the UrlRequest.Builder */
+    private void addBodyIfExists(@Nullable byte[] body, UrlRequest.Builder builder) {
+        if (body != null) {
+            UploadDataProvider dataProvider = UploadDataProviders.create(body);
+            builder.setUploadDataProvider(dataProvider, mCallbackExecutor);
+        }
+    }
+
+    private void setRequestPriority(Request<?> request, UrlRequest.Builder builder) {
+        builder.setPriority(request.getPriority().ordinal() + 1);
```

Comment:
this code strikes me as risky; nothing guarantees that we wouldn't add a new priority value or reorder them in some form, unlikely though it may be.

it's also not obvious to me that we should map in this manner, just because the order is relatively the same. In particular, we map "NORMAL" in Volley to "LOW" in Cronet, which is counterintuitive, and if you're sharing the Cronet queue with stuff outside Volley, would result in all of Volley's seemingly normal requests being deprioritized. I think it would be more intuitive to map NORMAL -> DEFAULT and combine HIGH/IMMEDIATE into REQUEST_PRIORITY_HIGHEST since Cronet doesn't offer two levels above DEFAULT.

I think it would be better to define a helper method like:

```java
private int getPriority(Request<?> request) {
  switch (request.getPriority()) {
    case LOW:
      return REQUEST_PRIORITY_LOW;
    case HIGH:
    case IMMEDIATE:
      return REQUEST_PRIORITY_HIGHEST;
    case MEDIUM:
    default:
      return REQUEST_PRIORITY_NORMAL;
    }
}
```

and call builder.setPriority(getPriority(request)); inline instead.

Reply:
this is definitely cleaner and safer. I'll make that change.